### PR TITLE
feat(grpo): add compute_metrics support to GRPOTrainer

### LINKED ISSUE #2959: `compute_metrics` in `GROPTrainer`
### Feature request

Can you please add `compute_metrics` in the `GROPTrainer`?

I can see its available in other, i.e., [DPOTrainer](https://huggingface.co/docs/trl/v0.15.2/dpo_trainer#trl.DPOTrainer.compute_metrics)

### Motivation

1. We can compute accuracy or downstream eval metrics over the evaluation dataset
2. That will help in more fine-grained logging
3. Better metric for `metric_for_best_model`

### Your contribution

I am open to submit a PR

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
GRPO lacked support for evaluation metrics independent of the training reward, and its prediction outputs prevented a simple callback passthrough. I added `compute_metrics` support that gathers completions and rewards across evaluation batches and calls the callback once per evaluation cycle. The targeted evaluation and callback tests pass, along with lint, format, and documentation style checks.
Fixes #2959.

---

> [!NOTE]
> **Medium Risk**
> Changes eval logging and buffering in the GRPO training loop; incorrect gathering or draining could skew metrics or checkpoint selection, but scope is limited to optional eval behavior.
> 
> **Overview**
> **GRPOTrainer** now accepts an optional **`compute_metrics`** callback so evaluation can log task metrics that are independent of the training reward (e.g. for **`metric_for_best_model`** when reward is a weak proxy).
> 
> Because GRPO’s **`prediction_step`** returns no predictions, the Trainer hook cannot run on its own. During eval, **`_generate_and_score_completions`** gathers generated completions and per-function reward scores across batches into buffers; **`log()`** builds an **`EvalPrediction`** (string completions in **`predictions`**, **`(N, R)` reward tensor in **`label_ids`**) and merges returned metrics as **`eval_<name>`**. Buffers are reset at the start of **`evaluation_loop`** so **`predict()`** does not contaminate the next eval.
> 
> New tests cover logging custom metrics, callback inputs, buffer isolation, **`compute_metrics=None`**, and coexistence with **`eval_reward`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c5edc86407914f3d8f595c210e50ce33cd3abc67. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by JulesRoussel2001** [comment]
Thanks for carrying this and for talking to Quentin, I am glad that the work is moving. Since the squash collapsed my commits into one under your authorship, would you mind adding a Co-authored-by: Jules Roussel <julesroussel.de.c@gmail.com> trailer to the commit message? Happy for you to keep driving the PR otherwise.

**Comment by behroozazarkhalili** [comment]
Done: the commit is now e2768fa8 with `Co-authored-by: Jules Roussel <julesroussel.de.c@gmail.com>`, same tree as before. On the squash: after the four closed PRs were reopened from my fork with the original commits cherry-picked, @qgallouedec asked that the agent-authored commits not be carried, so each branch was collapsed to a single commit of mine with the credit kept in the PR body. The trailer is the right way to keep your name on the merged history, so it stays.

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2958,6 +2990,24 @@ def _generate_and_score_completions(
...
+            # computes the same custom metrics everywhere and the resulting eval_<name> keys are present in
+            # the logs of every process rather than only the main one.
+            all_completions = gather_object(completions)
+            self._completions_for_compute_metrics.extend(all_completions)
+            if self._rewards_for_compute_metrics is None:
+                self._rewards_for_compute_metrics = rewards_per_func.cpu()
+            else:
+                self._rewards_for_compute_metrics = torch.cat(
+                    [self._rewards_for_compute_metrics, rewards_per_func.cpu()], dim=0
+                )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:3009]: Confirmed, and fixed in d566931c.

`Trainer.predict()` runs the evaluation loop and then only fires the predict callback, never `log()`, which is the sole place that drains these buffers. So a predict run leaves them populated and the next `evaluate()` concatenates its generations.

Measured on the PR head with the fix removed: an evaluation that should see 6 completions saw 12 after an intervening `predict()` call, `assert 12 == 6`. With the fix, 5 passed across the `compute_metrics` selection.

The reset lives in an `evaluation_loop` override rather than in `predict`, so both entry points are covered in one place. `test_predict_does_not_leak_compute_metrics_buffers` pins it: evaluate, predict, evaluate, and the two evaluations must report the same row count.
